pageserver/src/tenant/timeline.rs

```diff
@@ -3148,10 +3159,28 @@ impl Timeline {
                         FlushLayerError::Other(_) | FlushLayerError::CreateImageLayersError(_),
                     ) => {
                         error!("could not flush frozen layer: {err:?}");
-                        break err;
+                        break err.map(|_| ());
                     }
                 }
             };
+
+            // If our layer flushes didn't carry disk_consistent_lsn up to the `to_lsn` advertised
+            // to us via layer_flush_start_rx, then advance it here.
+            //
+            // This path is only taken for tenants with multiple shards: single sharded tenants should
+            // never encounter a gap in the wal.
+            debug_assert!(self.shard_identity.count.count() > 1 || flushed_to_lsn >= frozen_to_lsn);
+            if flushed_to_lsn < frozen_to_lsn && self.shard_identity.count.count() > 1 {
+                let old_disk_consistent_lsn = self.disk_consistent_lsn.load();
```

Comment:
I see this is a pre-existing pattern that we do `.load()` followed by `.store()`, but it happens only in one other location, so we now add two locations. If both can run in parallel, there is the potential for race conditions here that can make `disk_consistent_lsn` go backwards in the worst case.

Reply:
It's safe in practice because we only write `self.disk_consistent_lsn` from the flush loop, but yeah, I don't like this much either.  It also doesn't really need to be an atomic at the machine level: it's just a memory location that's written by one task and read by others.

I've refactored this into a `set_disk_consistent_lsn` helper that has a comment explaining the safety rules.